extensions/azdata/src/test/azdata.test.ts

```diff
@@ -296,3 +144,165 @@ describe('azdata', function () {
 		});
 	});
 });
+
+async function testLinuxUnsuccessfulUpgrade() {
+	const executeSudoCommandStub = sinon.stub(childProcess, 'executeSudoCommand').rejects();
+	const upgradePromise = azdata.checkAndUpdateAzdata(oldAzdata, outputChannelMock.object);
+	await should(upgradePromise).be.rejected();
+	should(executeSudoCommandStub.calledOnce).be.true();
+}
+
+async function testDarwinUnsuccessfulUpgrade() {
+	const executeCommandStub = sinon.stub(childProcess, 'executeCommand').rejects();
+	const upgradePromise = azdata.checkAndUpdateAzdata(oldAzdata, outputChannelMock.object);
+	await should(upgradePromise).be.rejected();
+	should(executeCommandStub.calledOnce).be.true();
+}
+
+async function testWin32UnsuccessfulUpgrade() {
+	sinon.stub(HttpClient, 'downloadFile').returns(Promise.resolve(__filename));
+	sinon.stub(childProcess, 'executeCommand').rejects();
+	const upgradePromise = azdata.checkAndUpdateAzdata(oldAzdata, outputChannelMock.object);
+	await should(upgradePromise).be.rejected();
+}
+
+async function testLinuxSuccessfulUpgrade(releaseJson: { win32: { version: string; link: string; }; darwin: { version: string; }; linux: { version: string; }; }) {
+	sinon.stub(HttpClient, 'getTextContent').returns(Promise.resolve(JSON.stringify(releaseJson)));
+	const executeCommandStub = sinon.stub(childProcess, 'executeCommand').callsFake(async (command: string, _args: string[]) => {
+		should(command).be.equal('add-apt-repository');
+		return { stdout: 'success', stderr: '' };
+	});
+	let callNumber = 0;
+	const executeSudoCommandStub = sinon.stub(childProcess, 'executeSudoCommand').callsFake(async (command: string) => {
+		callNumber++;
+		switch (callNumber) {
+			case 3:
+				should(command).match('/^curl/');
```

Comment:
I think this level of testing is too much and not providing any value. Just verifying that it was called at least once is enough - since we're not actually calling the methods on the platform then we should just rely on the implementer knowing what commands need to be executed to handle this correctly.

And FYI - if this was something you needed to actually do you can just use the onCall functions to do this instead of keeping counters yourself.

`stub.onCall(3).calledWith(...)`

Reply:
ok, I will keep only the most important verifications instead of verification of each call. I will user onCall. Thanks for the suggestion.

Reply:
done.